File: app/src/main/java/com/mdgd/pokemon/ui/pokemons/state/PokemonsScreenEffect.java
package com.mdgd.pokemon.ui.pokemons.state;

import com.mdgd.mvi.states.AbstractEffect;
import com.mdgd.pokemon.ui.pokemons.PokemonsContract;

public class PokemonsScreenEffect extends AbstractEffect<PokemonsContract.View> {

    public static class ShowDetails extends PokemonsScreenEffect {

        private final Long id;

        public ShowDetails(Long id) {
            this.id = id;
        }

        @Override
        protected void handle(PokemonsContract.View screen) {
            screen.proceedToNextScreen(id);
        }
    }


    public static class ShowErrorEffect extends PokemonsScreenEffect {

        private final Throwable error;

        public ShowErrorEffect(Throwable e) {
            error = e;
        }

        @Override
        protected void handle(PokemonsContract.View screen) {
            screen.showError(error);
        }
    }

}

File: app/src/main/java/com/mdgd/pokemon/ui/pokemons/state/PokemonsScreenEffect.kt
package com.mdgd.pokemon.ui.pokemons.state

import com.mdgd.mvi.states.AbstractEffect
import com.mdgd.pokemon.ui.pokemons.PokemonsContract

sealed class PokemonsScreenEffect : AbstractEffect<PokemonsContract.View>() {

    class Error(val error: Throwable?) : PokemonsScreenEffect() {

        override fun handle(screen: PokemonsContract.View) {
            screen.setProgressVisibility(false)
            screen.showError(error)
        }
    }

    class ShowDetails(val id: Long?) : PokemonsScreenEffect() {

        override fun handle(screen: PokemonsContract.View) {
            screen.setProgressVisibility(false)
            screen.proceedToNextScreen(id)
        }
    }

    class ScrollToStart : PokemonsScreenEffect() {

        override fun handle(screen: PokemonsContract.View) {
            screen.scrollToStart()
        }
    }
}
